### PR TITLE
Kernel: Add USB mouse support

### PR DESCRIPTION
**Pull request by supercomputer7**
To test this, add the following parameters to `run.sh`:
```
-M pc,i8042=off
-machine vmport=off
-device usb-mouse
```

Derived from #18673. We can get this merged quite quickly and later on figure out the details for HID keyboards, including USB ones.

cc @FireFox317

### CODE REVIEW

File: Kernel/Bus/USB/USBDevice.h [hunk 1]
```diff
@@ -75,6 +77,7 @@ class Device : public AtomicRefCounted<Device> {
 
     NonnullLockRefPtr<USBController> m_controller;
     NonnullOwnPtr<ControlPipe> m_default_pipe; // Default communication pipe (endpoint0) used during enumeration
+    OwnPtr<Pipe> m_interrupt_in_pipe;
```
Comment (FireFox317) [inline Kernel/Bus/USB/USBDevice.h:80]: Please store the pipe in the `USBMouseDevice`, instead of adding code to `USBDevice`

File: Kernel/Bus/USB/USBConfiguration.cpp [hunk 2]
```diff
@@ -50,11 +54,39 @@ ErrorOr<void> USBConfiguration::enumerate_interfaces()
...
+                }
+
+                // FIXME: Allow to set different configurations...
+                USBConfigurationDescriptor descriptor;
+                AK::TypedTransfer<USBConfigurationDescriptor>::copy(&descriptor, &m_descriptor, 1);
+                transfer_length = TRY(m_device.control_transfer(USB_REQUEST_TRANSFER_DIRECTION_DEVICE_TO_HOST, USB_REQUEST_SET_CONFIGURATION, 1, 0, sizeof(USBConfigurationDescriptor), &descriptor));
+                dbgln_if(USB_DEBUG, "USB device configuration was set!");
+
+                // FIXME: Detect USB HID mouse in an abstracted way...
+                if (interface_descriptor->interface_sub_class_code == 0x1 && interface_descriptor->interface_protocol == 0x2) {
```
Comment (FireFox317) [inline Kernel/Bus/USB/USBConfiguration.cpp:76]: This code shouldn't be in here, but in `USBPipe` instead. See the USB MassStorage PR

File: Kernel/Devices/HID/USB/MouseDevice.cpp [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    auto device = TRY(DeviceManagement::try_create_device<USBMouseDevice>(usb_device));
+    TRY(device->create_and_start_polling_process());
+    return *device;
+}
+
+ErrorOr<void> USBMouseDevice::create_and_start_polling_process()
+{
+    VERIFY(m_attached_usb_device->interrupt_pipe());
+    [[maybe_unused]] auto interrupt_in_transfer = TRY(static_cast<USB::InterruptInPipe*>(m_attached_usb_device->interrupt_pipe())->submit_interrupt_in_transfer(6, 10, [this](auto* transfer) {
+        if (transfer->transfer_data_size() == 0)
```
Comment (FireFox317) [inline Kernel/Devices/HID/USB/MouseDevice.cpp:29]: Should we check for `transfer->complete()` instead? Or we could also argue that it doesn't make sense to call the callback when we receive NAKs.

File: Kernel/Devices/HID/USB/MouseDevice.h [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+#include <AK/RefPtr.h>
+#include <Kernel/API/MousePacket.h>
+#include <Kernel/Bus/USB/USBDevice.h>
+#include <Kernel/Devices/HID/MouseDevice.h>
+#include <Kernel/Library/KString.h>
+#include <Kernel/Security/Random.h>
+
+namespace Kernel {
+
+class USBMouseDevice final : public MouseDevice {
```
Comment (FireFox317) [inline Kernel/Devices/HID/USB/MouseDevice.h:21]: Should this inherit from `USBDevice` instead? See the USB MassStorage PR. Or we could argue it should inherit from both classes I guess. Not really sure what the pattern is throughout the code

**Review by FireFox317 — CHANGES_REQUESTED** [review]

File: Kernel/Devices/HID/USB/MouseDevice.h (hunk 4 above)
Comment (supercomputer7) [reply Kernel/Devices/HID/USB/MouseDevice.h:21]: I dislike that pattern so I will not apply it here - what @Hendiadyoin1 did was essentially a hack with a singleton, due to a old limitation that we have with current code that any device should be controlled by a controller. What we really want in the future is to allow a device to be controlled directly be a controller or a `SOME_BUS::Driver` struct.

File: Kernel/Bus/USB/USBConfiguration.cpp (hunk 2 above)
Comment (supercomputer7) [reply Kernel/Bus/USB/USBConfiguration.cpp:76]: I don't see how @Hendiadyoin1 code did anything to change `USB::Pipe` to actually detect USB devices?

File: Kernel/Devices/HID/USB/MouseDevice.cpp (hunk 3 above)
Comment (supercomputer7) [reply Kernel/Devices/HID/USB/MouseDevice.cpp:29]: We probably shouldn't send transfers if we got a NAK in my opinion. @Quaker762 what do you think about it?

File: Kernel/Bus/USB/USBConfiguration.cpp (hunk 2 above)
Comment (Hendiadyoin1) [reply Kernel/Bus/USB/USBConfiguration.cpp:76]: IIRC the only thing I did to Pipes was make the interface nicer

File: Kernel/Bus/USB/USBConfiguration.cpp [hunk 5]
```diff
@@ -50,11 +54,38 @@ ErrorOr<void> USBConfiguration::enumerate_interfaces()
...
```
Comment (Hendiadyoin1) [inline Kernel/Bus/USB/USBConfiguration.cpp:76]: Can we make this not use magic numbers, but some kind of enum or other constant

**Comment by supercomputer7** [comment]
Relies on #21090 until it is merged, so this is a draft now.

File: Kernel/Bus/USB/Drivers/HID/MouseDriver.cpp [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+        //        although that's very unlikely
+        bool has_accepted_an_interface = false;
+        for (auto const& interface : config.interfaces()) {
+            // FIXME: Handle multiple interfaces
+            //        Interface may coexist at the same time,
+            //        but having multiple handles on the same data storage seems like a bad idea,
+            //        so:
+            // FIXME: Choose the best supported interface
+            //        UAS for example is supposed to be better than BBB, but BBB will always
+            //        be the first listed interface of them, when both are supported
```
Comment (Hendiadyoin1) [inline Kernel/Bus/USB/Drivers/HID/MouseDriver.cpp:57]: This seems a bit wrong for HID

File: Kernel/Bus/USB/Drivers/HID/MouseDriver.cpp [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+    HIDManagement::the().attach_standalone_hid_device(*mouse_device);
+    m_interfaces.append(mouse_device);
+    return {};
+}
+
+void MouseDriver::detach(USB::Device& device)
+{
+    auto&& mouse_device = AK::find_if(m_interfaces.begin(), m_interfaces.end(), [&device](auto& interface) { return &interface.device() == &device; });
+
+    HIDManagement::the().attach_standalone_hid_device(*mouse_device);
```
Comment (Hendiadyoin1) [inline Kernel/Bus/USB/Drivers/HID/MouseDriver.cpp:88]: I think this should say `detach_standalone_hid_device`

File: Kernel/Bus/USB/USBConfiguration.cpp [hunk 8]
```diff
@@ -5,10 +5,13 @@
  */
 
 #include <AK/FixedArray.h>
+#include <AK/TypedTransfer.h>
 #include <Kernel/Bus/USB/USBClasses.h>
 #include <Kernel/Bus/USB/USBConfiguration.h>
+#include <Kernel/Bus/USB/USBController.h>
 #include <Kernel/Bus/USB/USBInterface.h>
 #include <Kernel/Bus/USB/USBRequest.h>
+#include <Kernel/Devices/HID/Management.h>
```
Comment (Hendiadyoin1) [inline Kernel/Bus/USB/USBConfiguration.cpp:14]: Seems like a few to many new includes

File: Kernel/Bus/USB/USBConfiguration.cpp [hunk 9]
```diff
@@ -55,7 +58,7 @@ ErrorOr<void> USBConfiguration::enumerate_interfaces()
                 raw_endpoint_descriptor_offset += sizeof(USBHIDDescriptor); // Skip the HID descriptor (this was worked out via buffer inspection)
 
             USBEndpointDescriptor endpoint_descriptor;
-            memcpy(&endpoint_descriptor, raw_endpoint_descriptor_offset, sizeof(USBEndpointDescriptor));
+            AK::TypedTransfer<u8>::copy(reinterpret_cast<u8*>(&endpoint_descriptor), raw_endpoint_descriptor_offset, sizeof(USBEndpointDescriptor));
```
Comment (Hendiadyoin1) [inline Kernel/Bus/USB/USBConfiguration.cpp:61]: If this is needed I'd put it in a separate commit with a nice explanation why this can't/shouldn't be a memcpy

File: Kernel/Bus/USB/USBDevice.cpp [hunk 10]
```diff
@@ -7,11 +7,13 @@
 #include <AK/OwnPtr.h>
 #include <AK/Types.h>
 #include <AK/Vector.h>
+#include <Kernel/Bus/USB/UHCI/UHCIController.h>
 #include <Kernel/Bus/USB/USBController.h>
 #include <Kernel/Bus/USB/USBDescriptors.h>
 #include <Kernel/Bus/USB/USBDevice.h>
 #include <Kernel/Bus/USB/USBManagement.h>
 #include <Kernel/Bus/USB/USBRequest.h>
+#include <Kernel/Devices/HID/Management.h>
```
Comment (Hendiadyoin1) [inline Kernel/Bus/USB/USBDevice.cpp:16]: See above

File: Kernel/Devices/HID/USB/MouseDevice.cpp [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+{
+    auto device = TRY(DeviceManagement::try_create_device<USBMouseDevice>(usb_device, move(pipe)));
+    TRY(device->create_and_start_polling_process());
+    return *device;
+}
+
+ErrorOr<void> USBMouseDevice::create_and_start_polling_process()
+{
+    [[maybe_unused]] auto interrupt_in_transfer = TRY(m_interrupt_in_pipe->submit_interrupt_in_transfer(6, 10, [this](auto* transfer) {
+        i8 packet_raw[6];
```
Comment (Hendiadyoin1) [inline Kernel/Devices/HID/USB/MouseDevice.cpp:29]: I'd personally use a small simple struct, which seems more sensible than magic indexing

File: Kernel/Bus/USB/USBConfiguration.cpp (hunk 9 above)
Comment (ADKaster) [reply Kernel/Bus/USB/USBConfiguration.cpp:61]: Please don't use TypedTransfer in normal code. it's meant as implementation details of AK containers

File: Kernel/Devices/HID/USB/MouseDevice.cpp (hunk 3 above)
Comment (supercomputer7) [reply Kernel/Devices/HID/USB/MouseDevice.cpp]: I think that if a transfer ends with 0 bytes being transferred, it's probably not meaningful to invoke the async callback, hence I added a commit to just discard such transfer.

**Review by Hendiadyoin1 — APPROVED** [review]

File: Kernel/Bus/USB/Drivers/HID/MouseDriver.cpp [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+    // and handle multiple interfaces for it.
+    if (device.configurations()[0].interfaces().size() != 1)
+        return ENOTSUP;
+
+    TRY(checkout_interface(device, device.configurations()[0].interfaces()[0]));
+
+    return ENOTSUP;
+}
+
+ErrorOr<void> MouseDriver::initialise_device(USB::Device& device, USBInterface const& interface)
```
Comment (FireFox317) [inline Kernel/Bus/USB/Drivers/HID/MouseDriver.cpp:57]: We normally use the US spelling: `initialize`

File: Kernel/Boot/CommandLine.cpp [hunk 13]
```diff
@@ -226,6 +226,16 @@ UNMAP_AFTER_INIT bool CommandLine::is_physical_networking_disabled() const
     return contains("disable_physical_networking"sv);
 }
 
+UNMAP_AFTER_INIT bool CommandLine::disable_ps2_controller() const
+{
+    return contains("disable_ps2_controller"sv);
+}
```
Comment (FireFox317) [inline Kernel/Boot/CommandLine.cpp:232]: This method is not documented nor used anywhere in the code.

File: Kernel/Bus/USB/Drivers/HID/Codes.h [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+
+#include <AK/StdLibExtraDetails.h>
+#include <AK/StringView.h>
+#include <AK/Types.h>
+
+namespace Kernel::USB::HID {
+
+enum class SubclassCode : u8 {
+    BootProtocol = 0x01,
+};
```
Comment (FireFox317) [inline Kernel/Bus/USB/Drivers/HID/Codes.h:17]: It would be nice if there was some spec link added somewhere in this code.

File: Kernel/Bus/USB/USBConfiguration.cpp [hunk 15]
```diff
@@ -5,6 +5,7 @@
  */
 
 #include <AK/FixedArray.h>
+#include <AK/TypedTransfer.h>
```
Comment (FireFox317) [inline Kernel/Bus/USB/USBConfiguration.cpp:8]: Do we need this?

File: Kernel/Bus/USB/Drivers/HID/Codes.h (hunk 14 above)
Comment (supercomputer7) [reply Kernel/Bus/USB/Drivers/HID/Codes.h:17]: I don't think this is crucial here. I could try to find the reference from a USB specification if it's really necessary.

**Comment by gmta** [comment]
The first thing I noticed while testing this, is that the mouse buttons are barely functional and that mouse position updates are very choppy. I'm running Serenity on a fairly standard Ubuntu x86_64 machine with x86_64 build and Qemu from our repository. Is there an underlying FIXME somewhere?

**Comment by supercomputer7** [comment]
> The first thing I noticed while testing this, is that the mouse buttons are barely functional and that mouse position updates are very choppy. I'm running Serenity on a fairly standard Ubuntu x86_64 machine with x86_64 build and Qemu from our repository. Is there an underlying FIXME somewhere?

I sent a fix now to solve the problem. It was that I hardcoded the max packet size in the async transfer creation to 6 bytes, but now I saw it is actually 4 bytes. So instead of hardcoding it I just simply let the driver to figure out the correct max packet size and send it to the function that actually creates the `USBMouseDevice` and perform the async transfer setup.

**Review by gmta — APPROVED** [review]
Works nicely now! Thanks!
